File: TRT_Dynamic_Input/Infer_Onnx.cpp
// ./Infer_Onnx -s ../model/Dynamics_InputNet.onnx ./Dynamics_InputNet.engine
// ./Infer_Onnx -d ./Dynamics_InputNet.engine

#include <iostream>
#include <fstream> 
#include <string>
#include <cassert>
#include <chrono> 
#include "NvInfer.h"
#include "NvOnnxParser.h"  

using namespace nvinfer1;

static const int total_size = 1 * 3 * 256 * 256; // 大小
static const int output_size = 1 * 256 * 256 * 256;

// 继承 ILogger 类型创建自己的Logger
class Logger : public nvinfer1::ILogger{
	virtual void log(Severity severity, const char* msg) noexcept override{
		if (severity != Severity::kINFO)
			std::cout << msg << std::endl;
	}
};

class Infer_onnx{
public:
    Infer_onnx(){} // 构造函数
    nvinfer1::ICudaEngine* onnx2engine(std::string onnx_path, std::string engine_path); // 基于onnx模型，序列化模型，返回序列化好的engine
    void deserialize_and_infer(std::string engine_path, float* input_data, float* output_data, int in_size); // 反序列化并推理
    void runInference(IExecutionContext &context, float *input, float *output); // 执行推理
private:
    Logger gLogger; // Logger对象
    float *input; // 存放输入数据
    float *output; // 存放输出数据
    int input_size; // 输入数据的字节大小
    int output_size; // 输出数据的字节大小
};

nvinfer1::ICudaEngine* Infer_onnx::onnx2engine(std::string onnx_path, std::string engine_path){
    // 创建builder
    nvinfer1::IBuilder* builder = nvinfer1::createInferBuilder(gLogger);

    // 利用builder创建network（空的network，后面需要不断添加算子）
    // 1U << static_cast<uint32_t>(NetworkDefinitionCreationFlag::kEXPLICIT_BATCH) 表示使用显式batchsize
    // 传递 0 时，例如builder->createNetworkV2(0U)表示使用隐式batchsize，隐式batchsize
    // 显式batchsize和隐式batchsize参考：https://zhuanlan.zhihu.com/p/547973146
    nvinfer1::INetworkDefinition* network = builder->createNetworkV2(1U << static_cast<uint32_t>(nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH));

    // 借助nvidia提供的nvonnxparser类创建onnx模型解析器
    auto parser = nvonnxparser::createParser(*network, gLogger);
    bool parser_state = parser->parseFromFile(onnx_path.c_str(), 2); // 解析模型，在创建解析器的时候传入了network，因此在解析过程中会自动填充network
    if(parser_state == false){ // 解析成功返回true，解析失败返回false
        std::cerr << "parser onnx model failed!" << std::endl;
        exit(1);
    }

    // 利用builder创建配置config
    nvinfer1::IBuilderConfig* config = builder->createBuilderConfig();
    // 设置config
    config->setMaxWorkspaceSize(16 * (1 << 20)); // 设置最大工作空间
    config->setFlag(nvinfer1::BuilderFlag::kGPU_FALLBACK); // 启用GPU回退模式
    config->setFlag(nvinfer1::BuilderFlag::kFP16); // 设置精度计算
	nvinfer1::IOptimizationProfile* profile = builder->createOptimizationProfile(); // 创建优化配置文件
    // Dims4(dynamic_batchsize, channel, dynamic_height, dynamic_width) // 使用的onnx模型设置了三个动态维度
	profile->setDimensions("input", nvinfer1::OptProfileSelector::kMIN, nvinfer1::Dims4(1, 3, 224, 224)); // 设置输入input的动态维度，最小值
	profile->setDimensions("input", nvinfer1::OptProfileSelector::kOPT, nvinfer1::Dims4(3, 3, 256, 256)); // 期望输入的最优值
	profile->setDimensions("input", nvinfer1::OptProfileSelector::kMAX, nvinfer1::Dims4(5, 3, 512, 512)); // 最大值
	config->addOptimizationProfile(profile); // 将优化配置加入到config

    // 使用填充好的network和设置好的config来创建engine
    nvinfer1::ICudaEngine* engine = builder->buildEngineWithConfig(*network, *config);

    // 保存engine
    std::ofstream p(engine_path, std::ios::binary);
    if(!p){ // 打开失败
        std::cerr << "could not open plan output file" << std::endl;
        exit(1);
    }
    // 打开成功
    nvinfer1::IHostMemory* modelStream = engine->serialize(); // 序列化创建的engine
    p.write(reinterpret_cast<const char*>(modelStream->data()), modelStream->size()); // 保存

    // 释放所有
	modelStream->destroy();
	engine->destroy();
	network->destroy();
	parser->destroy();

    return engine; // 返回创建好的engine
}

void Infer_onnx::deserialize_and_infer(std::string engine_path, float* input_data, float* output_data, int in_size){
    this->input_size = in_size;
    this->output_size = this->input_size / 3 * 256;
    this->input = input_data;
    this->output = output_data;

    // 准备将模型写入 stream 中
    char *trtModelStream{nullptr};
    std::ifstream file(engine_path, std::ios::binary); // 读取 engine
    size_t engine_size{0}; // 初始化size大小
    if(file.good()){
        file.seekg(0, file.end); // 从file.end开始，偏移0
        engine_size = file.tellg(); // 获取当前位置，由于上一步处于file.end，因此可以获取file的字节大小
        file.seekg(0, file.beg); // 从file.beg开始，偏移0
        trtModelStream = new char[engine_size];
        assert(trtModelStream);
        file.read(trtModelStream, engine_size); // 读取size个size_t大小存放到创建的trtModelStream中
        file.close();
    }

    IRuntime *runtime = createInferRuntime(gLogger); // 创建runtime
    assert(runtime != nullptr);

    ICudaEngine *engine = runtime->deserializeCudaEngine(trtModelStream, engine_size, nullptr); // 反序列化加载engine
    assert(engine != nullptr);

    IExecutionContext *context = engine->createExecutionContext(); // 创建context
    context->setBindingDimensions(0, Dims4(1, 3, 256, 256)); // 这步非常重要，要为动态维度设置具体的数值，否则输出结果会是全零
    assert(context != nullptr);

    auto start = std::chrono::system_clock::now(); // 记录推理时长
    runInference(*context, this->input, this->output); // 执行推理
    auto end = std::chrono::system_clock::now(); // 打印推理时间
    std::cout << "\n[INFO]: Time taken by execution: "
              << std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() 
              << "ms" << std::endl;

    // 释放内存和空间
    context->destroy();
    engine->destroy();
    runtime->destroy();
}

void Infer_onnx::runInference(IExecutionContext &context, float *input, float *output){
    const ICudaEngine &engine = context.getEngine(); // 从context中获取engine
    assert(engine.getNbBindings() == 2); // 输入和输出两个bindings
    void *buffers[2]; // 初始化两个buffer用于拷贝输入数据和输出结果

    // 通过name来获取输入和输出binding的索引
    const int inputIndex = engine.getBindingIndex("input"); // input对应输入节点的名称，参考最初的onnx模型可视化结果
    const int outputIndex = engine.getBindingIndex("output"); // output对应输出节点的名称

    // 在GPU中为输入和输出申请内存
    cudaMalloc(&buffers[inputIndex], this->input_size * sizeof(float));
    cudaMalloc(&buffers[outputIndex],  this->output_size * sizeof(float));

    // 创建cudastream
    cudaStream_t stream;
    cudaStreamCreate(&stream);

    // 从 cpu 中拷贝输入到 gpu
    cudaMemcpyAsync(buffers[inputIndex], input, this->input_size * sizeof(float), cudaMemcpyHostToDevice, stream);
    // enqueue支持隐式batchsize, enqueueV2支持显式batchsize
    // https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_execution_context.html#a63cd95430852038ce864e17c670e0b36
    context.enqueueV2(buffers, stream, nullptr); // 执行推理

    // 将推理结果从 gpu 拷贝回 cpu 中
    cudaMemcpyAsync(output, buffers[outputIndex], this->output_size * sizeof(float), cudaMemcpyDeviceToHost, stream);

    cudaStreamSynchronize(stream); // 同步流，确保所有操作完成
    
    // 释放内存
    cudaStreamDestroy(stream);
    cudaFree(buffers[inputIndex]);
    cudaFree(buffers[outputIndex]);
}

int main(int argc, char* argv[]){
    Infer_onnx Infer_demo; // 实例化类对象
    if(argc == 4 && std::string(argv[1]) == "-s"){ // 序列化
        std::string onnx_path = argv[2];
        std::string engine_path = argv[3];
        Infer_demo.onnx2engine(onnx_path, engine_path); // 执行成员函数 onnx2engine
        std::cout << "serialize successfully" << std::endl;
    }
    else if(argc == 3 && std::string(argv[1]) == "-d"){
        std::string engine_path = argv[2];
        // 初始化测试数据
        static float input_data[total_size];
        for(int i = 0; i < total_size; i++){
            input_data[i] = 1.0;
        }

        // 初始化输出缓冲区
        static float output_data[output_size]; // 测试engine是将3通道原始特征变为256通道的高级特征
        // 序列化并推理
        Infer_demo.deserialize_and_infer(engine_path, input_data, output_data, total_size);
      
        // 打印第一个数据，验证推理结果，可以与python版本的对照来验证
        for(int i = 0; i < 1; i++){
            std::cout << output_data[i] << std::endl; 
        }
        std::cout << "deserialize successfully" << std::endl;
    }
    else{
        std::cerr << "input wrong" << std::endl;
        exit(1);
    }
    return 0;
}

File: TRT_Dynamic_Input/Infer_Onnx.py
import pycuda
import pycuda.driver as cuda
import pycuda.autoinit
import tensorrt as trt
import torch
import argparse
import numpy as np

def get_parser():
    parser = argparse.ArgumentParser(description = 'Parameters of Export Onnx Demo') 
    parser.add_argument(
        '--b',
        type = int,
        default = 2)
    parser.add_argument(
        '--c',
        type = int,
        default = 3)
    parser.add_argument(
        '--h',
        type = int,
        default = 256)
    parser.add_argument(
        '--w',
        type = int,
        default = 256)
    parser.add_argument(
        '--onnx',
        type = str,
        default = './model/Dynamics_InputNet.onnx')
    return parser

# 前处理
def preprocess(data):
    data = np.asarray(data)
    return data
    
# 后处理
def postprocess(data):
    data = np.reshape(data, (B, 256, H, W))
    return data
    
# 创建build_engine类
class build_engine():
    def __init__(self, onnx_path):
        super(build_engine, self).__init__()
        self.onnx = onnx_path
        self.engine = self.onnx2engine() # 调用 onnx2engine 函数生成 engine
        
    def onnx2engine(self):
        # 创建日志记录器
        TRT_LOGGER = trt.Logger(trt.Logger.WARNING)
        
        # 显式batch_size，batch_size有显式和隐式之分
        EXPLICIT_BATCH = 1 << (int)(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
        
        # 创建builder，用于创建network
        builder = trt.Builder(TRT_LOGGER)
        network = builder.create_network(EXPLICIT_BATCH) # 创建network（初始为空）
        
        # 创建config
        config = builder.create_builder_config()
        profile = builder.create_optimization_profile() # 创建profile
        profile.set_shape("input", (1,3,128,128), (3,3,256,256), (5,3,512,512))  # 设置动态输入,分别对应:最小尺寸、最佳尺寸、最大尺寸
        config.add_optimization_profile(profile)
        config.max_workspace_size = 1<<30 # 允许TensorRT使用1GB的GPU内存，<<表示左移，左移30位即扩大2^30倍，使用2^30 bytes即 1 GB
        
        # 创建parser用于解析模型
        parser = trt.OnnxParser(network, TRT_LOGGER)
        
        # 读取并解析模型
        onnx_model_file = self.onnx # Onnx模型的地址
        model = open(onnx_model_file, 'rb')
        if not parser.parse(model.read()): # 解析模型
            for error in range(parser.num_errors):
                print(parser.get_error(error)) # 打印错误（如果解析失败，根据打印的错误进行Debug）
 
        # 创建序列化engine
        engine = builder.build_serialized_network(network, config)
        # # 保存序列化的模型
        # if True:
        #     saved_trt_path = "./serialize_Dynamics.trt" # 序列化模型保存的地址
        #     with open(saved_trt_path, "wb") as f:
        #         f.write(engine) # 保存序列化模型
        return engine
        
    def get_engine(self):
        return self.engine # 返回 engine
    
# 分配内存缓冲区
def Allocate_memory(engine, context):
    bindings = []
    for binding in engine:
        binding_idx = engine.get_binding_index(binding) # 遍历获取对应的索引
        
        size = trt.volume(context.get_binding_shape(binding_idx))
        # context.get_binding_shape(binding_idx): 获取对应索引的Shape，例如input的Shape为(1, 3, H, W)
        # trt.volume(shape): 根据shape计算分配内存 
        
        dtype = trt.nptype(engine.get_binding_dtype(binding))
        # engine.get_binding_dtype(binding): 获取对应index或name的类型
        # trt.nptype(): 映射到numpy类型
        
        if engine.binding_is_input(binding): # 当前index为网络的输入input
            input_buffer = np.ascontiguousarray(input_data) # 将内存不连续存储的数组转换为内存连续存储的数组，运行速度更快
            input_memory = cuda.mem_alloc(input_data.nbytes) # cuda.mem_alloc()申请内存
            bindings.append(int(input_memory))
        else:
            output_buffer = cuda.pagelocked_empty(size, dtype)
            output_memory = cuda.mem_alloc(output_buffer.nbytes)
            bindings.append(int(output_memory))
            
    return input_buffer, input_memory, output_buffer, output_memory, bindings
 
        
if __name__ == "__main__":
    parser = get_parser()
    args = parser.parse_args()
    
    # 设置输入参数，生成输入数据
    Batch_size = args.b
    Channel = args.c
    Height = args.h
    Width = args.w
    input_data = torch.ones((Batch_size, Channel, Height, Width))
    
    # 前处理
    input_data = preprocess(input_data)
    
    # 生成engine
    onnx_model_file = args.onnx
    engine_build = build_engine(onnx_model_file)
    engine = engine_build.get_engine()
    
    # 生成context
    TRT_LOGGER = trt.Logger(trt.Logger.WARNING)
    runtime = trt.Runtime(TRT_LOGGER)
    engine = runtime.deserialize_cuda_engine(engine)
    context = engine.create_execution_context()
    
    # 绑定上下文
    B, C, H, W = input_data.shape
    context.set_binding_shape(engine.get_binding_index("input"), (B, 3, H, W))
    
    # 分配内存缓冲区
    input_buffer, input_memory, output_buffer, output_memory, bindings = Allocate_memory(engine, context)
    
    # 创建Cuda流
    stream = cuda.Stream()
    # 拷贝数据到GPU （host -> device)
    cuda.memcpy_htod_async(input_memory, input_buffer, stream) # 异步拷贝数据
    
    # 推理
    context.execute_async_v2(bindings=bindings, stream_handle=stream.handle)
    
    # 将GPU得到的推理结果 拷贝到主机（device -> host）
    cuda.memcpy_dtoh_async(output_buffer, output_memory, stream)
    
    # 同步Cuda流
    stream.synchronize()
    
    # 后处理
    print("Test Result: ", output_buffer[0]) # 对比CPP版本
    output_data = postprocess(output_buffer)
    print("output.shape is : ", output_data.shape)
